pkg/automatic_policy/automatic_policy.go

```diff
@@ -0,0 +1,116 @@
+	Title string
+	// Extension is the extension of the software package.
+	Extension string
+	// BundleIdentifier contains the bundle identifier for 'pkg' packages.
+	BundleIdentifier string
+	// PackageIDs contains the product code for 'msi' packages.
+	PackageIDs []string
+}
+
+var (
+	// ErrExtensionNotSupported is returned if the extension is not supported to generate automatic policies.
+	ErrExtensionNotSupported = errors.New("extension not supported")
+	// ErrMissingBundleIdentifier is returned if the software extension is "pkg" and a bundle identifier was not extracted from the installer.
+	ErrMissingBundleIdentifier = errors.New("missing bundle identifier")
+	// ErrMissingProductCode is returned if the software extension is "msi" and a product code was not extracted from the installer.
+	ErrMissingProductCode = errors.New("missing product code")
+	// ErrMissingTitle is returned if a title was not extracted from the installer.
+	ErrMissingTitle = errors.New("missing product code")
+)
+
+// Generate generates the "trigger policy" from the metadata of a software package.
+func Generate(metadata InstallerMetadata) (*PolicyData, error) {
+	switch {
+	case metadata.Title == "":
+		return nil, ErrMissingTitle
+	case metadata.Extension != "pkg" && metadata.Extension != "msi" && metadata.Extension != "deb" && metadata.Extension != "rpm":
+		return nil, ErrExtensionNotSupported
+	case metadata.Extension == "pkg" && metadata.BundleIdentifier == "":
+		return nil, ErrMissingBundleIdentifier
+	case metadata.Extension == "msi" && (len(metadata.PackageIDs) == 0 || metadata.PackageIDs[0] == ""):
+		return nil, ErrMissingProductCode
+	}
+
+	name := fmt.Sprintf("[Install software] %s (%s)", metadata.Title, metadata.Extension)
+
+	description := fmt.Sprintf("Policy triggers automatic install of %s on each host that's missing this software.", metadata.Title)
+	if metadata.Extension == "deb" || metadata.Extension == "rpm" {
+		basedPrefix := "Debian"
+		if metadata.Extension == "rpm" {
+			basedPrefix = "RPM"
+		}
+		description += fmt.Sprintf(
+			"\nSoftware won't be installed on Linux hosts with %s-based distributions because this policy's query is written to always pass on these hosts.",
+			basedPrefix,
+		)
+	}
+
+	switch metadata.Extension {
+	case "pkg":
+		return &PolicyData{
+			Name:        name,
+			Query:       fmt.Sprintf("SELECT 1 FROM apps WHERE bundle_identifier = '%s';", metadata.BundleIdentifier),
+			Platform:    "darwin",
+			Description: description,
+		}, nil
+	case "msi":
+		return &PolicyData{
+			Name:        name,
+			Query:       fmt.Sprintf("SELECT 1 FROM programs WHERE identifying_number = '%s';", metadata.PackageIDs[0]),
+			Platform:    "windows",
+			Description: description,
+		}, nil
+	case "deb":
+		return &PolicyData{
+			Name: name,
+			Query: fmt.Sprintf(
+				// First inner SELECT will mark the policies as successful on RHEL hosts.
+				`SELECT 1 WHERE EXISTS (
+	SELECT 1 FROM os_version WHERE platform = 'rhel'
```

Comment:
Good idea. Though it would break if users are installing rpm packages on Debian-based systems, or viceversa (deb packages on RPM-based systems). Via `alien`, though not sure how common this is...

Reply:
Yeah, I think we can assume folks aren't using `alien` here, and folks who are can tweak policies to their liking anyway.

Reply:
[Add better query to detect RPM-based vs DEB-based systems](https://github.com/fleetdm/fleet/pull/25021/commits/b5ed87d9491e03b228c425d548da4099fc23ac25)

Will do proper QA tomorrow, but this seems to work.

Reply:
Tests on VMs are passing, so this is ready for re-review.